### PR TITLE
Cleanup: HIR ty lowering: Consolidate the places that do assoc item probing & access checking

### PR DESCRIPTION
**Pull request by fmease**
Use `probe_assoc_item` (for hygienically probing an assoc item and checking if it's accessible wrt. visibility and stability) for assoc item constraints, too, not just for assoc type paths and make the privacy error translatable.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs [hunk 1]
```diff
@@ -428,15 +412,15 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             });
 
             // Provide the resolved type of the associated constant to `type_of(AnonConst)`.
-            if let hir::AssocItemConstraintKind::Equality { term: hir::Term::Const(anon_const) } =
-                constraint.kind
-            {
-                let ty = alias_ty.map_bound(|ty| tcx.type_of(ty.def_id).instantiate(tcx, ty.args));
-                let ty = check_assoc_const_binding_type(tcx, assoc_ident, ty, constraint.hir_id);
+            if let Some(anon_const) = constraint.ct() {
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:415]: unrelated drive-by cleanup

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs [hunk 2]
```diff
@@ -428,15 +412,15 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                let ty =
+                    alias_term.map_bound(|ty| tcx.type_of(ty.def_id).instantiate(tcx, ty.args));
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:417]:
```suggestion
                    alias_term.map_bound(|alias| tcx.type_of(alias.def_id).instantiate(tcx, alias.args));
```

**Review by davidtwco — APPROVED** [review]

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs [hunk 3]
```diff
@@ -292,30 +292,15 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
-        tcx.check_stability(assoc_item.def_id, Some(hir_ref_id), constraint.span, None);
+        let assoc_item = self
+            .probe_assoc_item(
+                constraint.ident,
+                assoc_kind,
+                hir_ref_id,
+                constraint.span,
+                candidate.def_id(),
+            )
+            .expect("failed to find associated item");
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:303]: I don't mind that name, feel free to do this in a follow-up.
